src/main/java/com/commercetools/sync/producttypes/helpers/AttributeDefinitionReferenceResolver.java

```diff
@@ -83,16 +85,54 @@ public AttributeDefinitionReferenceResolver(
       final SetAttributeType setAttributeType = (SetAttributeType) attributeType;
       final AttributeType elementType = setAttributeType.getElementType();
 
-      if (elementType instanceof NestedAttributeType) {
+      final AtomicInteger maxDepth = new AtomicInteger();
+      AttributeType nestedAttributeType = elementType;
 
-        return resolveNestedTypeReference((NestedAttributeType) elementType)
-            .thenApply(SetAttributeType::of)
-            .thenApply(attributeDefinitionDraftBuilder::attributeType);
+      if (elementType instanceof SetAttributeType) {
+        maxDepth.incrementAndGet();
+        nestedAttributeType = getNestedAttributeType((SetAttributeType) elementType, maxDepth);
+      }
+      if (nestedAttributeType instanceof NestedAttributeType) {
+        return resolveNestedAttributeTypeReferences(
+            attributeDefinitionDraftBuilder, maxDepth, (NestedAttributeType) nestedAttributeType);
       }
     }
     return completedFuture(attributeDefinitionDraftBuilder);
   }
 
+  @Nonnull
+  private CompletionStage<AttributeDefinitionDraftBuilder> resolveNestedAttributeTypeReferences(
```

Comment:
A little comment would help as this is pretty complex.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/720/commits/4eaf98b9f722d6c2b4614192dadd9a29b6037ee3